### PR TITLE
remove account state serialization and access account data directly f…

### PR DESCRIPTION
**Pull request by bflood366**
…or reads/writes

### CODE REVIEW

**Review by bholzman — APPROVED** [review]
Looks good!

File: arch/contracts/exchange/program/src/lib.rs [hunk 1]
```diff
@@ -61,29 +62,233 @@ pub enum NetworkType {
...
+
+    pub fn set_num_balances(account: &AccountInfo, num_balances: usize) -> Result<(), ProgramError> {
+        let offset = TOKEN_STATE_HEADER_SIZE - 4;
+        let mut data = account.data.try_borrow_mut().map_err(|_| ProgramError::InvalidAccountData)?;
+        Ok(data[offset..offset+4].copy_from_slice((num_balances as u32).to_le_bytes().as_slice()))
+    }
+
+    pub fn get_token_id(account: &AccountInfo) -> Result<String, ProgramError> {
+        let mut tmp = [0u8; MAX_TOKEN_ID_LENGTH];
+        let offset = MAX_TOKEN_ID_LENGTH + 36;
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs]: I cleaned up all the hard coded values and replaces with constants

File: arch/contracts/exchange/program/src/lib.rs [hunk 2]
```diff
@@ -61,29 +62,233 @@ pub enum NetworkType {
...
+        serialized.extend((self.balances.len() as u32).to_le_bytes());
+        for balance in self.balances.iter() {
+            serialized.extend(&balance.to_vec());
+        }
+
+        serialized
+    }
+
+    pub fn get_num_balances(account: &AccountInfo) -> Result<usize, ProgramError> {
+        let offset = TOKEN_STATE_HEADER_SIZE - 4;
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs]: added a constant here

File: arch/contracts/exchange/program/src/lib.rs [hunk 3]
```diff
@@ -61,29 +62,233 @@ pub enum NetworkType {
...
+        if index as usize >= accounts.len() {
+            return Err(ProgramError::Custom(ERROR_INVALID_ACCOUNT_INDEX));
+        }
+        if TokenBalances::get_program_state_account_key(&accounts[index as usize])? != *accounts[0].key {
+            return Err(ProgramError::Custom(ERROR_PROGRAM_STATE_MISMATCH));
+        }
+        Ok(())
+    }
+
+    fn grow_account_if_needed(account: &AccountInfo, additional_len: usize) -> Result<(), ProgramError> {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs]: changed to pass in the number of balance records

File: arch/contracts/exchange/program/src/lib.rs [hunk 4]
```diff
@@ -462,43 +606,48 @@ fn verify_decrements(state: &TokenBalances, adjustments: Vec<Adjustment>) -> Res
...
                 Some(new_balance) => new_balance,
                 None => return Err(ProgramError::Custom(ERROR_INSUFFICIENT_BALANCE))
             };
             if withdrawal.fee_amount > 0 {
-                if state.balances[FEE_ADDRESS_INDEX as usize].address != fee_account_address {
+                if Balance::get_wallet_address(account, FEE_ADDRESS_INDEX as usize)? != fee_account_address {
                     return Err(ProgramError::Custom(ERROR_ADDRESS_MISMATCH))
                 }
-                state.balances[FEE_ADDRESS_INDEX as usize].balance += withdrawal.fee_amount
+                Balance::set_wallet_balance(account, FEE_ADDRESS_INDEX as usize, Balance::get_wallet_balance(account, FEE_ADDRESS_INDEX as usize)? + withdrawal.fee_amount)?
```
[the remark this answers is not in the document]
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs]: added a new method here per your suggestion

File: arch/contracts/exchange/program/src/lib.rs [hunk 5]
```diff
@@ -93,7 +317,7 @@ pub struct Adjustment {
 }
 
 #[derive(Clone, BorshSerialize, BorshDeserialize)]
-pub struct TokenBalanceSetup {
+pub struct TokenStateetup {
```
Comment (antonyshchenko) [inline arch/contracts/exchange/program/src/lib.rs:320]: Typo in struct name

File: arch/contracts/exchange/program/src/lib.rs [hunk 6]
```diff
@@ -307,18 +514,20 @@ pub fn withdraw_batch(program_id: &Pubkey, accounts: &[AccountInfo], params: Wit
...
 
-pub fn submit_settlement_batch(accounts: &[AccountInfo], params: SettlementBatchParams) -> Result<(), ProgramError> {
-    let mut program_state: ProgramState = get_account_state(accounts, 0)?;
+pub fn submit_settlement_batch(accounts: &[AccountInfo], params: &SettlementBatchParams) -> Result<(), ProgramError> {
+
+    let current_hash = ProgramState::get_settlement_hash(&accounts[0])?;
+    let params_hash = hash(borsh::to_vec(&params).unwrap());
 
-    if program_state.settlement_batch_hash.is_empty() {
+    if current_hash == [0u8; 32] {
```
Comment (antonyshchenko) [inline arch/contracts/exchange/program/src/lib.rs:522]: I've noticed `[0u8; 32]` being used in several places. Should we define a constant for it perhaps?

File: arch/contracts/exchange/src/lib.rs [hunk 7]
```diff
@@ -53,29 +53,52 @@ mod tests {
...
     pub struct ProgramState {
-        pub version: u16,
+        pub version: u32,
         pub fee_account_address: String,
         pub program_change_address: String,
         pub network_type: NetworkType,
-        pub settlement_batch_hash: String,
-        pub last_settlement_batch_hash: String,
-        pub last_withdrawal_batch_hash: String,
+        pub settlement_batch_hash: [u8; 32],
```
Comment (antonyshchenko) [inline arch/contracts/exchange/src/lib.rs:99]: Is there a way to make a type-alias for `[u8; 32]` ?

File: arch/contracts/exchange/program/src/lib.rs (hunk 5 above)
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs:320]: I just opened another PR where I fixed this.

File: arch/contracts/exchange/program/src/lib.rs (hunk 6 above)
Comment (bflood366) [reply arch/contracts/exchange/program/src/lib.rs:522]: I created a const for `EMPTY_HASH` in the other PR

**Pull request merged**